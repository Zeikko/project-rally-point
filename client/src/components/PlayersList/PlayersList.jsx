import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styled from 'styled-components'
import { getPlayersAction } from '../../actions/player-actions'
import * as propTypes from '../../constants/prop-types'

class PlayersList extends Component {
  componentDidMount() {
    const { dispatch, game } = this.props
    dispatch(getPlayersAction(game.data.id))
  }

  render() {
    const { players } = this.props
    if (!players.data) {
      return null
    }
    return (
      <div>
        <Heading>Players waiting for a game</Heading>
        {players.data.map(player => (
          <div key={player.id}>
            <img alt={player.displayName} src={player.smallAvatarUrl} />
            {player.displayName} ({player.country})
          </div>
        ))}
      </div>
    )
  }
}

PlayersList.propTypes = {
  dispatch: PropTypes.func.isRequired,
  game: propTypes.game.isRequired,
  players: propTypes.players.isRequired,
}

export default PlayersList

const Heading = styled.div`
  font-size: 20px;
`
