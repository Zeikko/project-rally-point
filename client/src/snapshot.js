import renderer from 'react-test-renderer'
import 'jest-styled-components'

export function matchComponentToSnapshot(component) {
  const json = renderer.create(component).toJSON()
  expect(json).toMatchSnapshot()
}