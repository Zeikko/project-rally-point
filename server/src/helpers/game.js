import _ from 'lodash'
import db from '../db'
import gameStatuses from '../../../common/game-statuses.json'
import playerRoles from '../../../common/player-roles.json'

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadLeaderPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfCaptainVotes = await db('captainVote')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfCaptainVotes[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadMemberPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfSquadLeaders = await db('player')
    .select('id')
    .where({
      gameId,
      role: playerRoles.SQUAD_LEADER,
      team: game.teamWithTurnToPick,
    })
  return (numberOfSquadLeaders.length + 1) === getOptimalNumberOfSquads(game)
}

export async function shouldStartPlayingGame(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
    .whereNotNull('team')
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.VOTE_CAPTAINS })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadLeaderPick(gameId) {
  const games = await db('game')
    .update({
      status: gameStatuses.SQUAD_LEADER_PICK,
      teamWithTurnToPick: 1,
    })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadMemberPick(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.SQUAD_MEMBER_PICK })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startPlayingGame(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.PLAYING })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

export async function passSquadLeaderPickTurn(game) {
  const games = await db('game')
    .update({
      teamWithTurnToPick: (game.teamWithTurnToPick % 2) + 1,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

export async function passPlayerPickTurn(game, playersCount) {
  const nextTeamWithTurnToPick = await calculatePickTurn(game, playersCount)
  const games = await db('game')
    .update({
      teamWithTurnToPick: nextTeamWithTurnToPick,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

async function calculatePickTurn(game, playersCount) {
  const numberOfSquads = getOptimalNumberOfSquads(game)
  if (playersCount % numberOfSquads === 0) {
    return (game.teamWithTurnToPick % 2) + 1
  }
  return game.teamWithTurnToPick
}

export function getOptimalNumberOfSquads(game) {
  return Math.ceil(game.maxPlayers / 2 / 9)
}
