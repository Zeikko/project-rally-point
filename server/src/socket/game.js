import _ from 'lodash'
import db from '../db'
import actions from '../../../common/actions.json'

export default function handleGame(io, socket, action) {
  switch (action.type) {
    case actions.GET_GAME_REQUEST:
      return getGameRequest(io, socket, action)
    default:
      return null
  }
}

export async function getGameRequest(io, socket) {
  try {
    const game = await db('game')
      .first('*')
      .where({
        status: 'queue',
      })
    socket.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
  } catch (error) {
    socket.emit('action', { type: actions.GET_GAME_ERROR })
  }
}

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: 'voting captains' })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function createNewGame() {
  return db('game').insert({})
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

