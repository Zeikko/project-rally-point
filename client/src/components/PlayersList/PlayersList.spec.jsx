import React from 'react'
import { matchComponentToSnapshot } from '../../test/snapshot'
import PlayersList from './PlayersList'
import initialState from '../../initial-state'
import normalUser from '../../../../fixtures/normal-user.json'
import game from '../../../../fixtures/game.json'

describe('PlayersList', () => {
  it('renders initial state correctly', () => {
    matchComponentToSnapshot(<PlayersList
      dispatch={() => {}}
      game={{ data: game }}
      players={{ data: []}}
    />)
  })

  it('renders a player correctly', () => {
    matchComponentToSnapshot(<PlayersList
      dispatch={() => {}}
      game={{ data: game }}
      players={{ data: [normalUser]}}
    />)
  })
})
