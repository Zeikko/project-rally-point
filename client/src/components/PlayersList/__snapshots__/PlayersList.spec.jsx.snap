// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayersList renders a player correctly 1`] = `
.c0 {
  font-size: 20px;
}

<div>
  <div
    className="c0"
  >
    Players waiting for a game
  </div>
  <div>
    <img
      alt="Test user"
      src="http://avatar.test/small.jpg"
    />
    Test user
     (
    FI
    )
  </div>
</div>
`;

exports[`PlayersList renders initial state correctly 1`] = `null`;
